packages/context/src/value-promise.ts

```diff
@@ -32,8 +32,8 @@ export type MapObject<T> = {[name: string]: T};
  *
  * @param value The value to check.
  */
-export function isPromise<T>(
-  value: T | PromiseLike<T>,
+export function isPromiseLike<T>(
```

Comment:
What's the expectation that we want to set for developers, `PromiseLike` or `Promise`? IIUC, `PromiseLike` does not have `catch` function. Is it going to be a surprise to developers that want to use compliant promises? Maybe we should switch from `PromiseLike` to `Promise`.

Reply:
For the record, `Promise.prototype.catch()` is available since ES2015.

Reply:
+1 for switching to `Promise` instead of us having to do https://github.com/strongloop/loopback-next/pull/1050/files#diff-f726e786caf6fb7f3616d8a58c63b3bfR58 (but that's the only instance I found). Not sure of other reasons.

Reply:
You are right that `Promise.prototype.catch` is available since ES2015. It's the other API, `Promise.prototype.finally` that is coming in ES2018 IIRC - see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally.

> What's the expectation that we want to set for developers, `PromiseLike` or `Promise`? 

My personal take based on [Robustness principle](https://en.wikipedia.org/wiki/Robustness_principle) (_Be conservative in what you do, be liberal in what you accept from others._):

 - When expecting a Promise created by the user/caller, we should allow any `PromiseLike` value. This simplifies integration with existing 3rd party libraries that do not use native promises yet, our own juggler is a prime example. 
 - When returning a Promise to the user/caller, always return a native Promise.

Please note that this is exactly how `async/await` works according to the spec - you can `await` any promise-like value, `async` function always return a native promise.

The changes proposed in this pull request are going in this direction.

> Is it going to be a surprise to developers that want to use compliant promises? 

IMO, as long as we always return native promises, developers should not face any surprise.

Having said that, I would be surprised if LoopBack did not accept non-native Promises (e.g. as returned by Bluebird) and forced me to convert non-native promises returned by 3rd party modules to native promises.